### PR TITLE
Fix OnCompleted and implement CompleteAsync

### LINKED ISSUE #17268: HttpResponse.OnCompleted isn't working as expected
I have tested this defect on .net core 2.2 and 3.0. It's a bug when using IIS Express but it's working as expected when switch to Kestrel. OnCompleted doesn't fire after response. 
Steps to reproduce:
Run the following code in IIS express and put the breakpoint on "await Task.Delay(2000); " line.
You will see the response is sent in the browser and then the following code is executed when using the Kestrel server but when you switch to IIS Express, the response is not sent in the browser.

```    
  [HttpGet]
        public  string Get()
        {
            try
            {
           
                // Some work 
                return "talal";
            }
            finally
            {

                Response.OnCompleted(async () =>
                {
                    // Put breakpoint here
                    await Task.Delay(2000); 
                });
            }
        }
```

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/dotnet/aspnetcore/issues/17268.

This introduces a behavior change to call Flush(moreData: false) when we complete the body output. By calling Flush(moreData: false), the response is sent to the client, but the HttpContext is still valid. 

I'm slightly concerned about perf now that we are calling another flush, but I also think perf could be better because we are complete the response earlier now.

I'd like to go through and add a bunch of tests that I skipped for CompleteAsync now (ex for trailers, reset, etc).

### CODE REVIEW

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs [hunk 1]
```diff
@@ -199,14 +199,26 @@ Task IHttpResponseBodyFeature.SendFileAsync(string path, long offset, long? coun
 
         // TODO: In the future this could complete the body all the way down to the server. For now it just ensures
         // any unflushed data gets flushed.
-        protected Task CompleteResponseBodyAsync()
+        protected async Task CompleteResponseBodyAsync()
         {
             if (ResponsePipeWrapper != null)
             {
-                return ResponsePipeWrapper.CompleteAsync().AsTask();
+                await ResponsePipeWrapper.CompleteAsync().AsTask();
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:206]: Pretty sure this isn't needed
```suggestion
                await ResponsePipeWrapper.CompleteAsync();
```

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs [hunk 2]
```diff
@@ -199,14 +199,26 @@ Task IHttpResponseBodyFeature.SendFileAsync(string path, long offset, long? coun
...
+            }
+
+            // Completing the body output will trigger a final flush to IIS.
+            // We'd rather not bypass the bodyoutput to flush, to guarantee we avoid
+            // calling flush twice at the same time.
+            // awaiting the writeBodyTask guarantees the response has finished the final flush.
+            _bodyOutput.Complete();
+            await _writeBodyTask;
+
+            return;
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:221]: nit: not needed

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs [hunk 3]
```diff
@@ -1315,11 +1318,113 @@ public async Task Reset_DuringRequestBody_Resets(HttpContext httpContext)
             }
         }
 
+        public Task Goaway(HttpContext httpContext)
+        {
+            httpContext.Response.Headers["Connection"] = "close";
+            return Task.CompletedTask;
+        }
+
+        private TaskCompletionSource<object> _completeAsync = new TaskCompletionSource<object>();
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1327]: nit: `TaskCompletionSource`

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResponseBodyTests.cs [hunk 4]
```diff
@@ -30,5 +30,31 @@ public async Task ResponseBodyTest_FlushedPipeAndThenUnflushedPipe_AutoFlushed()
...
+        }
+
+        [ConditionalFact]
+        [RequiresNewHandler]
+        public async Task ResponseBodyTest_CompleteAsyncWorks()
+        {
+            var response = await _fixture.Client.GetAsync("/CompleteAsync");
+            Assert.True(response.IsSuccessStatusCode);
+            var response2 = await _fixture.Client.GetAsync("/CompleteAsync_Completed");
+            Assert.True(response2.IsSuccessStatusCode);
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResponseBodyTests.cs:56]: Comment on what is going on here. Its not obvious to the reader (me 😋) without looking at the server impl

Applies to all the tests that have multiple calls that depend on each other

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs [hunk 5]
```diff
@@ -1315,11 +1318,139 @@ public async Task Reset_DuringRequestBody_Resets(HttpContext httpContext)
             }
         }
 
+        public Task Goaway(HttpContext httpContext)
+        {
+            httpContext.Response.Headers["Connection"] = "close";
+            return Task.CompletedTask;
+        }
+
+        private TaskCompletionSource _completeAsync = new TaskCompletionSource();
```
Comment (halter73) [inline src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1327]: I see you're living dangerously not using TaskCreationOptions.RunContinuationsAsynchronously.

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs [hunk 6]
```diff
@@ -199,14 +199,24 @@ Task IHttpResponseBodyFeature.SendFileAsync(string path, long offset, long? coun
 
         // TODO: In the future this could complete the body all the way down to the server. For now it just ensures
         // any unflushed data gets flushed.
-        protected Task CompleteResponseBodyAsync()
+        protected async Task CompleteResponseBodyAsync()
```
Comment (halter73) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:202]: Is this called every request? Would it be worth it to fast-path this when all the returned (Value)Tasks are completed?

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs [hunk 7]
```diff
@@ -199,22 +199,51 @@ Task IHttpResponseBodyFeature.SendFileAsync(string path, long offset, long? coun
...
         // TODO: In the future this could complete the body all the way down to the server. For now it just ensures
         // any unflushed data gets flushed.
-        protected async Task CompleteResponseBodyAsync()
+        protected Task CompleteResponseBodyAsync()
         {
             if (ResponsePipeWrapper != null)
             {
-                await ResponsePipeWrapper.CompleteAsync();
+                var completeAsyncTask = ResponsePipeWrapper.CompleteAsync();
+                if (!completeAsyncTask.IsCompletedSuccessfully)
```
Comment (JamesNK) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs:207]: If the ValueTask is completed I believe you still want to call `.GetAwaiter().GetResult();`, e.g.

https://github.com/dotnet/aspnetcore/blob/3509323ad182f47ea19cc20f5cfd5735eb447720/src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs#L18-L19

File: src/Servers/IIS/IIS/src/Core/IISHttpContextOfT.cs [hunk 8]
```diff
@@ -26,32 +26,37 @@ public IISHttpContextOfT(MemoryPool<byte> memoryPool, IHttpApplication<TContext>
...
+                    ReportApplicationError(ex);
+                    success = false;
+                }
+
+                if (ResponsePipeWrapper != null)
+                {
+                    await ResponsePipeWrapper.CompleteAsync();
+                }
+
+                _streams.Stop(); // check what this does.
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/src/Core/IISHttpContextOfT.cs:59]: ?

**Review by Tratcher — COMMENTED** [review]
The code looks good. Why are the tests failing?

File: src/Servers/IIS/IIS/src/Core/IISHttpContextOfT.cs (hunk 8 above)
Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpContextOfT.cs:59]: I'm looking into why tests are failing now. Somehow shutdown isn't happening correctly.

**Comment by jkotalik** [comment]
Rebasing fixed this for some reason. I was able to repro this every time locally as well; I believe the fixed I made to tests here: https://github.com/dotnet/aspnetcore/commit/56be06bc7d610c2ebf26424ca2b95d2cda5b445f#diff-0fbfbafe55296d34cf1554647120b2b6R62 fixed it.

Going to run CI again on this change to make sure.

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResponseBodyTests.cs [hunk 9]
```diff
@@ -30,5 +30,33 @@ public async Task ResponseBodyTest_FlushedPipeAndThenUnflushedPipe_AutoFlushed()
         {
             Assert.Equal(20, (await _fixture.Client.GetByteArrayAsync($"/FlushedPipeAndThenUnflushedPipe")).Length);
         }
+
+        //[ConditionalFact]
+        //[RequiresNewHandler]
+        //public async Task ResponseBodyTest_BodyCompletionNotBlockedByOnCompleted()
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResponseBodyTests.cs:36]: ?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**